### ISSUE TITLE
Survey tool: Custom Handler Plugin

### ISSUE THREAD

**Issue by beamalsky** [datamade/just-spaces#25]
We'll need a way to connect data collected by the fobi surveys forms with the Just Spaces models, which (will) follow the structure implied by the PLDP.

I'll build a custom fobi form handler plugin handler: https://django-fobi.readthedocs.io/en/0.13.8/#creating-a-new-form-handler-plugin

**Comment by beamalsky** [comment]
I spent part of yesterday getting acquainted with Fobi data handling and[ modifying the package's form creation function so that it automatically attaches a default data saving method to new forms](https://github.com/datamade/django-fobi/pull/1/commits/0e161712232c0afbc0cd3d9ca8d44193bdc4c96b). Now that I better understand the Fobi defaults, I'd like to zoom out, get my thoughts in order, and talk about app-wide data modeling before I proceed.

## Data Handling, two paths forward 

### Option 1: Use the Fobi method of storing data and translate it into the Just Spaces models later

The Fobi handler has one model for saving form data: `SavedFormDataEntry`. It has these fields:

- `id`
- `user`: the user who submitted the form
- `form_data_headers`: a stringified JSON object with each form element `{'form_element_name':'form_element_label'}`
- `saved_data`: a stringified JSON object with each form element `{'form_element_name':'saved_data'}`
- `created`: date and time form was submitted
- `form_entry`: foreign key of the form that was submitted

This captures the information we need if we assume that a form will not change once it's been submitted (which I think we are, but we'd have to be very careful!). It plays nicely with Fobi's built in export views. On the other hand, those JSON blobs are clunky and far from the data structure described by the PLDP, and I fear they would make Part 2 of this project harder than necessary.

### Option 2: Modify the Fobi data handler to store more information about elements and line up with the PLDP
The PLDP has separate models for Survey and SurveyComponent, which we could employ here. 

Survey (corresponds with Fobi's SavedFormDataEntry)
- `id`
- `user`
- `created`
- `form_entry`

SurveyComponent
- `id`
- `survey`: foreign key
- `component_name`
- `component_label`
- `component_uid`: the type of component, eg boolean or pldp_age
- `position`: refers to question order in the survey
- `saved_data`

This should be something that's usable directly for the analyses in part 2 of Just Spaces, and is a more normalized and less JSON-y structure. It means that surveys could be edited and contextual information about entered data won't be lost. It would likely also mean more work to use other parts of the Fobi package, as it expects data to be structured a certain way.

@reginafcompton I'm going back and forth on the relative benefits here, and I'd love to discuss!

**Comment by reginafcompton** [comment]
@beamalsky - thank you for this thoughtful overview. My inclinations say, "Option 2!", but I think we should do some time estimates, before leaping into action:

**For Option 1**, it's already implemented! That's great. However, as you say, "those JSON blobs are clunky." _Can we think a little more about how these JSON blobs would slow us down?_ 

One major pitfall that springs to mind: the JSON metadata could be difficult to query. [Psql does enable querying on JSONB fields.](https://hackernoon.com/how-to-query-jsonb-beginner-sheet-cheat-4da3aa5082a3) But I fear we might have difficulties querying the data across multiple surveys. Putting data into a `SurveyComponent` field (option 2) would ameliorate my apprehensions. Does that sound right to you? Perhaps, you've encountered complex use cases with psql and JSON? (Then again, there's a reason some dev teams use [non-relational databases](https://aws.amazon.com/nosql/)....)

**For Option 2**, you mention "It would likely also mean more work to use other parts of the Fobi package." _Can we determine what we need to customize on the Fobi side and/or build ourselves, e.g., the export functionality? and then estimate how many hours would be required?_ 

In short, we need to determine **which option will cost more hours**. It's easier to make estimates for Option 2, as opposed to Option 1 (since Part 2 has so many unknowns). How about you start with estimations for Option 2, e.g., put together a mini-mini-scope? Then, let's touch base.  

***

Unrelated! But an important note:
```
a form will not change once it's been submitted (which I think we are, but we'd have to be very careful!)
```
Yes, I agree, and I think we need to put restrictions in place to insure it, e.g., flagging forms that have been "run" and, if so, turning off edit capabilities. We can think about this more later on, but I think we can at least open an issue to flag it.

**Comment by beamalsky** [comment]
It sounds like we should lock in the idea that survey forms will not be editable once they're run, which is great! Would you like me to open that issue?

**On the JSON blobs in Option 1:**

- I definitely agree that they will likely make more complex queries more difficult
- I'm also concerned that not all the information we'll need later is being saved in a stable way (`component_uid` and `position` specifically, as well as making sure that we're collecting `row` data in a way that works with the PLDP, though I think that should be handled on the form creation side of things). Committing to not editing surveys after they're run would allay this concern.

**Things that need to be rewritten in Fobi, if we were to use Option 2**

- Data exporting
- All Fobi views to look at collected data

I think we'll be writing custom views for the collected data anyway, but the Fobi default export options would be useful. These are the only Fobi functions I'm certain a new handler would break, but it's possible there are others I'm not anticipating and this option does come with more risk.

As we've talked about this, it's occurred to me that there might be a third option: a data processing step (similar to a import_data command) that would move collected survey data from Fobi's data structures into one that matches the PLDP. That would mean that I won't be fighting the package in the way it saves data, but would also mean that collected data is moved into a format in line with PLDP and ready to be used by Part 2. I'll put together a time estimate for option 2, but in the meantime what do you think of that?

**Comment by beamalsky** [comment]
I believe it would take about 8 more hours, all told, to implement Option 2 and get csv exports from Fobi working again. That's not accounting for custom collected data views or unexpected problems with the package. 

I'd like to talk more with you about when a `process_collected_data` command would run if you think that idea is feasible, but I estimate it would take about 5 hours to implement.

**Comment by reginafcompton** [comment]
My sundry responses:

(1) "It sounds like we should lock in the idea that survey forms will not be editable once they're run, which is great! Would you like me to open that issue?"

Yes, please do!

(2) I am not thrilled about a `process_collected_data` script. It adds a layer of complexity and fragility that I think we can avoid, and more importantly, we'd need to figure out how often to run it against the data base – ideally, it would run as often as possible, but even then, every survey would have a period, in which the data only exists in it's JSON-ified state. Any automatic refreshing of data on the public-facing site or collected data views would be reliant on a cron successfully running this script. 

Does that seem like a fair assessment? 

(3) We have 80 hours scoped for "Analysing, accessing, and exporting survey data" (3.3), and at least part of what you describe can be billed against this. I think that seems reasonable. I also wonder why we need to use the Fobi data exports at all? There's loads of good models for exporting data from a database into a CSV file (including Large Lots). 

Would implementing our own functionality for the export component be less time consuming, than fighting against Fobi? 

***

Let me know your answers to the above, and then, we can make an informed decision.

**Comment by beamalsky** [comment]
That's an excellent point that we have a separate scope item for accessing and exporting survey data, I'd forgotten! That changes this a bit, and makes me feel better about option 2 because I'll have some more troubleshooting time in case it breaks Fobi in unexpected ways. 

I think we've definitely reached a level of departure from Fobi's intended use where yes, it would be easier to write our own export scripts than to try to use the package's.

**Comment by beamalsky** [comment]
New issue for publishing states: https://github.com/datamade/just-spaces/issues/54

**Comment by reginafcompton** [comment]
@beamalsky – I think option 2 makes the most sense, but I'd be happy to let this sit with us for a couple weeks....just to be sure.

**Comment by beamalsky** [comment]
I agree!

**Comment by beamalsky** [comment]
We're moving forward on the plan of modifying Fobi's data handling to match the needs of Just Spaces/the PLDP! 

Unfortunately, Fobi's [instructions to create a custom handler](https://django-fobi.readthedocs.io/en/0.13.8/#creating-a-new-form-handler-plugin) do not work. I followed their example code to the letter, then spent about 2 hours yesterday modifying it and checking registrations and imports. No dice. I then spent some time modifying the default data handler in our `django-fobi` fork, which works just fine. The potential downside there is that it needs PLDP models imported, which feels hacky—@regina could I get your input on whether or not this method is viable?

**Comment by reginafcompton** [comment]
@beamalsky - absolutely! Could you push up what you already have? That will give me some context.

**Comment by beamalsky** [comment]
Completed by #64. @jeancochrane, let's check in on how you'd like to manage closing issues!
